source/mir/random/flex/package.d

```diff
@@ -0,0 +1,925 @@
+    S l1 = f1(points[0]);
+    S l2 = f2(points[0]);
+
+    version(Flex_logging)
+    {
+        import std.experimental.logger;
+        log("starting flex with p=", points, ", cs=", cs);
+    }
+
+    // initialize with user given splitting points
+    foreach (i, r; points[1..$])
+    {
+        // reuse computations
+        S r0 = f0(r);
+        S r1 = f1(r);
+        S r2 = f2(r);
+        auto iv = transformToInterval(l, r, cs[i], l0, l1, l2,
+                                                   r0, r1, r2);
+        l = r;
+        l0 = r0;
+        l1 = r1;
+        l2 = r2;
+
+        calcInterval(iv);
+        totalHatAreaSummator += iv.hatArea;
+        totalSqueezeAreaSummator += iv.squeezeArea;
+
+        ips.insertBack(iv);
+    }
+
+    version(Flex_logging)
+    {
+        import std.algorithm;
+        import std.array;
+        log("----");
+        log("Interval: ", ips.array.map!`a.lx`);
+        log("hatArea", ips.array.map!`a.hatArea`);
+        log("squeezeArea", ips.array.map!`a.squeezeArea`);
+        log("----");
+    }
+
+    // Flex is not guaranteed to converge
+    for (auto i = 0; i < maxIterations && nrIntervals < apprMaxPoints; i++)
+    {
+        immutable totalHatArea = totalHatAreaSummator.sum;
+        immutable totalSqueezeArea = totalSqueezeAreaSummator.sum;
+
+        // Flex aims for a user defined efficiency
+        if (totalHatArea / totalSqueezeArea <= rho)
+            break;
+
+        version(Flex_logging)
+        {
+            tracef("iteration %d: totalHat: %.3f, totalSqueeze: %.3f, rho: %.3f",
+                    i, totalHatArea, totalSqueezeArea, totalHatArea / totalSqueezeArea);
+        }
+
+        immutable avgArea = nextDown(totalHatArea - totalSqueezeArea) / nrIntervals;
+        for (auto it = ips[]; !it.empty;)
+        {
+            immutable curArea = it.front.hatArea - it.front.squeezeArea;
+            if (curArea > avgArea)
+            {
+                // prepare total areas for update
+                totalHatAreaSummator -= it.front.hatArea;
+                totalSqueezeAreaSummator -= it.front.squeezeArea;
+
+                // split the interval at the arcmean into two parts
+                auto mid = arcmean!S(it.front);
+
+                // cache
+                immutable c = it.front.c;
+
+                // calculate new values
+                S mx0 = f0(mid);
+                S mx1 = f1(mid);
+                S mx2 = f2(mid);
+
+                // apply transformation to right side (for c=0 no transformations are applied)
+                S mt0 = (c == 0) ? mx0 : copysign(S(1), c) * exp(c * mx0);
+                S mt1 = (c == 0) ? mx1 : c * mt0 * mx1;
+                S mt2 = (c == 0) ? mx2 : c * mt0 * (c * pow(mx1, 2) + mx2);
+
+                Interval!S midIP = Interval!S(mid, it.front.rx, it.front.c,
+                                              mt0, mt1, mt2,
+                                              it.front.rtx, it.front.rt1x, it.front.rt2x);
```

Comment:
@9il I inlined as requested, but won't this make it super hard to experiment with other , custom transformations?